File: JobApplicationTracker/frontend/src/types/job.ts
export type CompensationType = 'hourly' | 'annual';

export type JobStatus =
  | 'applied'
  | 'interviewing'
  | 'offer_received'
  | 'accepted'
  | 'rejected'
  | 'withdrawn';

export interface JobApplication {
  id: number;
  job_title: string;
  company: string;
  compensation_min: number;
  compensation_max?: number;
  compensation_type: CompensationType;
  application_date: string;
  job_posting_url: string;
  resume_filename: string;
  cover_letter_filename?: string;
  status: JobStatus;
  got_interview: boolean;
  coding_question_link?: string;
  created_at: string;
  updated_at: string;
}

export interface JobApplicationCreate {
  job_title: string;
  company: string;
  compensation_min: number;
  compensation_max?: number;
  compensation_type: CompensationType;
  application_date: string;
  job_posting_url: string;
  resume_filename: string;
  cover_letter_filename?: string;
}

export interface JobApplicationUpdate {
  status?: JobStatus;
  got_interview?: boolean;
  coding_question_link?: string;
}

export interface JobFormData {
  jobTitle: string;
  company: string;
  compensationMin: string;
  compensationMax: string;
  compensationType: CompensationType;
  applicationDate: string;
  jobPostingUrl: string;
  resumeFilename: string;
  coverLetterFilename: string;
}

export interface UploadedFile {
  id: number;
  filename: string;
  file_type: 'resume' | 'cover_letter';
  original_filename: string;
  content_type: string;
  file_size: number;
  uploaded_at: string;
}

export interface StatusCount {
  status: JobStatus;
  count: number;
}

export interface ApplicationOverTime {
  date: string;
  count: number;
}

export interface RecentApplication {
  id: number;
  job_title: string;
  company: string;
  status: JobStatus;
  application_date: string;
}

export interface DashboardData {
  total_applications: number;
  applications_by_status: StatusCount[];
  interview_rate: number;
  applications_over_time: ApplicationOverTime[];
  recent_applications: RecentApplication[];
}

File: JobApplicationTracker/backend/app/models/job.py
from __future__ import annotations

import enum
from datetime import date, datetime
from typing import Optional

from sqlalchemy import String, Integer, Date, DateTime, Boolean, Enum
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class CompensationType(str, enum.Enum):
    HOURLY = "hourly"
    ANNUAL = "annual"


class JobStatus(str, enum.Enum):
    APPLIED = "applied"
    INTERVIEWING = "interviewing"
    OFFER_RECEIVED = "offer_received"
    ACCEPTED = "accepted"
    REJECTED = "rejected"
    WITHDRAWN = "withdrawn"


class JobApplication(Base):
    __tablename__ = "job_applications"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)

    # Fields set on creation
    job_title: Mapped[str] = mapped_column(String(100), nullable=False)
    company: Mapped[str] = mapped_column(String(50), nullable=False)
    compensation_min: Mapped[int] = mapped_column(Integer, nullable=False)
    compensation_max: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    compensation_type: Mapped[CompensationType] = mapped_column(
        Enum(CompensationType), nullable=False
    )
    application_date: Mapped[date] = mapped_column(Date, nullable=False)
    job_posting_url: Mapped[str] = mapped_column(String, nullable=False)
    resume_filename: Mapped[str] = mapped_column(String, nullable=False)
    cover_letter_filename: Mapped[Optional[str]] = mapped_column(String, nullable=True)

    # Fields set post-creation (inline edit)
    status: Mapped[JobStatus] = mapped_column(
        Enum(JobStatus), nullable=False, default=JobStatus.APPLIED
    )
    got_interview: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    coding_question_link: Mapped[Optional[str]] = mapped_column(String, nullable=True)

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=datetime.utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow
    )
